base/applications/rapps/include/asyncinet.h

```diff
@@ -0,0 +1,68 @@
+#ifndef ASYNC_INET
+#define ASYNC_INET
+
+#include <windows.h>
+#include <wininet.h>
+
+enum ASYNC_EVENT
+{
+    ASYNCINET_DATA,           // wParam is the Data retrieved from the internet, lParam is the length of Data
+
+    ASYNCINET_COMPLETE,       // wParam and lParam is not used.
+                              // when receiving this, AsyncInet will be free soon and should not used anymore
+
+    ASYNCINET_CANCELLED,      // wParam and lParam is not used.
+                              // when receiving this, AsyncInet will be free soon and should not used anymore
+
+    ASYNCINET_ERROR           // wParam is not used. lParam specify the error code (if there is one). 
+                              // when receiving this, AsyncInet will be free soon and should not used anymore
+};
+
+typedef struct __AsyncInet ASYNCINET, * pASYNCINET;
+
+typedef int
+(*ASYNCINET_CALLBACK)(
+    pASYNCINET AsyncInet,
+    ASYNC_EVENT Event,
+    WPARAM wParam,
+    LPARAM lParam,
+    VOID* Extension
+    );
+
+typedef struct __AsyncInet
+{
+    HINTERNET hInternet;
+    HINTERNET hInetFile;
+
+    HANDLE hEventHandleCreated;
+
+    UINT ReferenceCnt;
+    CRITICAL_SECTION CriticalSection;
+    HANDLE hEventRefZero; // TODO: Use conditional variable instead when one day ROS has conditional variable
+    HANDLE hEventHandleClose;
+
+    BOOL bIsOpenUrlComplete;
+
+    BOOL bCleanUp;
```

Comment:
bCleanup is probably not needed,
when decrementing a reference count to 0, cleanup is implied (and noone else should have a reference to this object anymore!)
a function / thread should not receive an object with reference count 0, you could even assert on that (since someone handed you a pointer that was not 'safe' anymore, since they forgot to take a reference.